src/transaction.rs

```diff
@@ -0,0 +1,254 @@
+};
+use ffi;
+use libc::{c_char, size_t};
+use std::marker::PhantomData;
+
+pub struct Transaction<'a> {
+    pub(crate) inner: *mut ffi::rocksdb_transaction_t,
+    pub(crate) db: PhantomData<&'a OptimisticTransactionDB>,
+    snapshot: Option<*const ffi::rocksdb_snapshot_t>,
+}
+
+impl<'a> Transaction<'a> {
+    pub(crate) fn new(
+        inner: *mut ffi::rocksdb_transaction_t,
+        snapshot: Option<*const ffi::rocksdb_snapshot_t>,
+    ) -> Transaction<'a> {
+        Transaction {
+            inner,
+            db: PhantomData,
+            snapshot,
+        }
+    }
+
+    /// commits a transaction
+    pub fn commit(&self) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_commit(self.inner,));
+        }
+        Ok(())
+    }
+
+    /// Delete a key inside a transaction
+    ///
+    /// ColumnFamilyHandle: default
+    pub fn delete(&self, key: &[u8]) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_delete(
+                self.inner,
+                key.as_ptr() as *const c_char,
+                key.len() as size_t,
+            ));
+            Ok(())
+        }
+    }
+
+    /// Delete a key inside a transaction
+    pub fn delete_cf(&self, cf: ColumnFamily, key: &[u8]) -> Result<(), Error> {
+        unsafe {
+            ffi_try!(ffi::rocksdb_transaction_delete_cf(
+                self.inner,
+                cf.inner,
+                key.as_ptr() as *const c_char,
+                key.len() as size_t,
+            ));
+            Ok(())
+        }
+    }
+
+    /// Read a key inside a transaction
+    ///
+    /// ColumnFamilyHandle: default
+    pub fn get_opt(
+        &self,
+        key: &[u8],
+        mut readopts: ReadOptions,
+    ) -> Result<Option<DBVector>, Error> {
+        if readopts.inner.is_null() {
+            return Err(Error::new(
+                "Unable to create RocksDB read options. \
+                 This is a fairly trivial call, and its \
+                 failure may be indicative of a \
+                 mis-compiled or mis-loaded RocksDB \
+                 library."
+                    .to_owned(),
+            ));
+        }
+
+        if let Some(snapshot) = self.snapshot {
+            readopts.set_snapshot(snapshot)
+        };
+
+        unsafe {
+            let mut val_len: size_t = 0;
+            let val = ffi_try!(ffi::rocksdb_transaction_get(
+                self.inner,
+                readopts.inner,
+                key.as_ptr() as *const c_char,
+                key.len() as size_t,
+                &mut val_len,
+            )) as *mut u8;
+            if val.is_null() {
+                Ok(None)
+            } else {
+                Ok(Some(DBVector::from_c(val, val_len)))
+            }
+        }
+    }
+
+    /// Read a key inside a transaction
+    ///
+    /// ReadOptions: default
+    /// ColumnFamilyHandle: default
+    pub fn get(&self, key: &[u8]) -> Result<Option<DBVector>, Error> {
```

Comment:
Can you make `key` generic? eg: `K: AsRef<[u8]>` See #253 for some examples of similar changes.